e2e_playwright/st_dialog_test.py

```diff
@@ -0,0 +1,114 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction, wait_for_app_run
+
+
+def test_displays_dialog_properly(app: Page):
+    """Test that dialog is displayed properly."""
+    app.get_by_text("Open Dialog").click()
+    wait_for_app_run(app)
+    main_dialog = app.locator("[data-testid='stModal']")
+    expect(main_dialog).to_have_count(1)
+
+
+def test_dialog_closes_properly(app: Page):
+    """Test that dialog closes after clicking on action button."""
+    app.get_by_text("Open Dialog").click()
+    wait_for_app_run(app)
+    main_dialog = app.locator("[data-testid='stModal']")
+    expect(main_dialog).to_have_count(1)
+    close_button = main_dialog.get_by_test_id("stButton").locator("button").first
+    close_button.scroll_into_view_if_needed()
+    close_button.click()
+    wait_for_app_run(app)
+    main_dialog = app.locator("[data-testid='stModal']")
+    expect(main_dialog).to_have_count(0)
```

Comment:
This is basically doing a fullscreen screenshot, which is a bit problematic because it can easily get flaky and also requires updates when unrelated things change. For the main menu modal tests, we are doing this screenshot tests just based on the dialog window, e.g.:

```suggestion
    expect(main_dialog).get_by_role("dialog")
```

Maybe also something to consider here.

Reply:
Great point, I will update it!

---

Though I guess you wanted to comment on the other test that actually takes a screenshot, right?

```python
def test_dialog_displays_correctly(app: Page, assert_snapshot: ImageCompareFunction):
    app.get_by_text("Open Dialog").click()
    wait_for_app_run(app)
    # assert_snapshot(app.locator(".main"), name="dialog-in-main")
    assert_snapshot(app.get_by_role("dialog"), name="dialog-in-main")
```

Reply:
ups, yes indeed 👍